`ClassUtil` fails with `java.lang.reflect.InaccessibleObjectException` trying to setAccessible on `OptionalInt` with JDK 17+

### Search before asking

- [X] I searched in the [issues](https://github.com/FasterXML/jackson-databind/issues) and found nothing similar.

### Describe the bug

Please consider the following trivial Java code:

```java
package org.myapp;

import com.fasterxml.jackson.databind.ObjectMapper;
import java.util.OptionalInt;

public class Main {
    public static void main(final String[] args) throws Exception {
        final ObjectMapper objectMapper = new ObjectMapper();
        final String json = "{ }"; // empty
        final Data data = objectMapper.readValue(json, Data.class);
        System.out.println("Read data: " + data);
    }

    public static class Data {
        private OptionalInt value;

        public Data() {

        }

        public void setValue(OptionalInt i) {
            this.value = i;
        }


        public OptionalInt getValue() {
            return this.value;
        }

        @Override
        public String toString() {
            return "Data[value=" + this.value + "]";
        }
    }
}
```
When using `jackson-databind` `2.15.2` and Java version 17 and running this program, it results in:

```console
Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make private java.util.OptionalInt() accessible: module java.base does not "opens java.util" to unnamed module @4cf328c3
    at java.lang.reflect.AccessibleObject.throwInaccessibleObjectException (AccessibleObject.java:387)
    at java.lang.reflect.AccessibleObject.checkCanSetAccessible (AccessibleObject.java:363)
    at java.lang.reflect.AccessibleObject.checkCanSetAccessible (AccessibleObject.java:311)
    at java.lang.reflect.Constructor.checkCanSetAccessible (Constructor.java:192)
    at java.lang.reflect.Constructor.setAccessible (Constructor.java:185)
    at com.fasterxml.jackson.databind.util.ClassUtil.checkAndFixAccess (ClassUtil.java:995)
    at com.fasterxml.jackson.databind.deser.impl.CreatorCollector._fixAccess (CreatorCollector.java:278)
    at com.fasterxml.jackson.databind.deser.impl.CreatorCollector.setDefaultCreator (CreatorCollector.java:130)
    at com.fasterxml.jackson.databind.deser.BasicDeserializerFactory._addExplicitConstructorCreators (BasicDeserializerFactory.java:438)
    at com.fasterxml.jackson.databind.deser.BasicDeserializerFactory._constructDefaultValueInstantiator (BasicDeserializerFactory.java:293)
    at com.fasterxml.jackson.databind.deser.BasicDeserializerFactory.findValueInstantiator (BasicDeserializerFactory.java:222)
    at com.fasterxml.jackson.databind.deser.BeanDeserializerFactory.buildBeanDeserializer (BeanDeserializerFactory.java:262)
    at com.fasterxml.jackson.databind.deser.BeanDeserializerFactory.createBeanDeserializer (BeanDeserializerFactory.java:151)
    at com.fasterxml.jackson.databind.deser.DeserializerCache._createDeserializer2 (DeserializerCache.java:415)
    at com.fasterxml.jackson.databind.deser.DeserializerCache._createDeserializer (DeserializerCache.java:350)
    at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCache2 (DeserializerCache.java:264)
    at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCacheValueDeserializer (DeserializerCache.java:244)
    at com.fasterxml.jackson.databind.deser.DeserializerCache.findValueDeserializer (DeserializerCache.java:142)
    at com.fasterxml.jackson.databind.DeserializationContext.findNonContextualValueDeserializer (DeserializationContext.java:644)
    at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.resolve (BeanDeserializerBase.java:539)
    at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCache2 (DeserializerCache.java:294)
    at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCacheValueDeserializer (DeserializerCache.java:244)
    at com.fasterxml.jackson.databind.deser.DeserializerCache.findValueDeserializer (DeserializerCache.java:142)
    at com.fasterxml.jackson.databind.DeserializationContext.findRootValueDeserializer (DeserializationContext.java:654)
    at com.fasterxml.jackson.databind.ObjectMapper._findRootDeserializer (ObjectMapper.java:4956)
    at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose (ObjectMapper.java:4826)
    at com.fasterxml.jackson.databind.ObjectMapper.readValue (ObjectMapper.java:3772)
    at com.fasterxml.jackson.databind.ObjectMapper.readValue (ObjectMapper.java:3740)
    at org.myapp.Main.main (Main.java:10)
```
So `com.fasterxml.jackson.databind.util.ClassUtil` is trying to `setAccessible()` on the private constructor of a JDK class `java.util.OptionalInt`. One way to solve this issue is to configure the `ObjectMapper` instance as follows:

```java
objectMapper.configure(MapperFeature.CAN_OVERRIDE_ACCESS_MODIFIERS, false);
```

However, while looking at the code in `com.fasterxml.jackson.databind.util.ClassUtil` I noticed that there's a specific logic which tries to not `setAccessible()` on JDK internal classes here https://github.com/FasterXML/jackson-databind/blob/jackson-databind-2.15.2/src/main/java/com/fasterxml/jackson/databind/util/ClassUtil.java#L994 which looks like:

```java
if (!isPublic || (evenIfAlreadyPublic && !isJDKClass(declaringClass))) {
      ao.setAccessible(true);
  }
```

Should that `!isJDKClass(declaringClass)` be perhaps applied even when `evenIfAlreadyPublic` is false? Something like:

```java
if (!isJDKClass(declaringClass) && (!isPublic || evenIfAlreadyPublic)) {
      ao.setAccessible(true);
  }
```
That way, it won't try to access the internal JDK classes and run into these exceptions on Java 17+?



### Version Information

2.15.2

### Reproduction

_No response_

### Expected behavior

_No response_

### Additional context

_No response_

--- a/release-notes/VERSION-2.x
+++ b/release-notes/VERSION-2.x
@@ -63,6 +63,8 @@ Project: jackson-databind
 #4078: `java.desktop` module is no longer optional
  (reported by Andreas Z)
  (fix contributed by Joo-Hyuk K)
+#4082: `ClassUtil` fails with `java.lang.reflect.InaccessibleObjectException`
+  trying to setAccessible on `OptionalInt` with JDK 17+
 
 2.15.3 (not yet released)
 

--- a/src/main/java/com/fasterxml/jackson/databind/util/BeanUtil.java
+++ b/src/main/java/com/fasterxml/jackson/databind/util/BeanUtil.java
@@ -309,6 +309,9 @@ public class BeanUtil
         } else if (isJodaTimeClass(className)) {
             typeName =  "Joda date/time";
             moduleName = "com.fasterxml.jackson.datatype:jackson-datatype-joda";
+        } else if (isJava8OptionalClass(className)) {
+            typeName =  "Java 8 optional";
+            moduleName = "com.fasterxml.jackson.datatype:jackson-datatype-jdk8";
         } else {
             return null;
         }
@@ -323,8 +326,21 @@ public class BeanUtil
         return isJava8TimeClass(rawType.getName());
     }
 
+    // @since 2.12
     private static boolean isJava8TimeClass(String className) {
         return className.startsWith("java.time.");
+    }
+
+    /**
+     * @since 2.16
+     */
+    public static boolean isJava8OptionalClass(Class<?> rawType) {
+        return isJava8OptionalClass(rawType.getName());
+    }
+
+    // @since 2.16
+    private static boolean isJava8OptionalClass(String className) {
+        return className.startsWith("java.util.Optional");
     }
 
     /**
@@ -334,6 +350,7 @@ public class BeanUtil
         return isJodaTimeClass(rawType.getName());
     }
 
+    // @since 2.12
     private static boolean isJodaTimeClass(String className) {
         return className.startsWith("org.joda.time.");
     }

--- a/src/test/java/com/fasterxml/jackson/databind/interop/OptionalJava8Fallbacks4082Test.java
+++ b/src/test/java/com/fasterxml/jackson/databind/interop/OptionalJava8Fallbacks4082Test.java
@@ -1,0 +1,89 @@
+package com.fasterxml.jackson.databind.interop;
+
+import java.util.Optional;
+import java.util.OptionalDouble;
+import java.util.OptionalInt;
+import java.util.OptionalLong;
+
+import com.fasterxml.jackson.core.*;
+
+import com.fasterxml.jackson.databind.*;
+import com.fasterxml.jackson.databind.exc.InvalidDefinitionException;
+import com.fasterxml.jackson.databind.util.TokenBuffer;
+
+// [databind#4082]: add fallback handling for Java 8 Optional types, to
+// prevent accidental serialization as POJOs, as well as give more information
+// on deserialization attempts
+//
+// @since 2.16
+public class OptionalJava8Fallbacks4082Test extends BaseMapTest
+{
+    private final ObjectMapper MAPPER = newJsonMapper();
+
+    // Test to prevent serialization as POJO, without Java 8 date/time module:
+    public void testPreventSerialization() throws Exception {
+        _testPreventSerialization(Optional.empty());
+        _testPreventSerialization(OptionalInt.of(13));
+        _testPreventSerialization(OptionalLong.of(-1L));
+        _testPreventSerialization(OptionalDouble.of(0.5));
+    }
+
+    private void _testPreventSerialization(Object value) throws Exception
+    {
+        try {
+            String json = MAPPER.writeValueAsString(value);
+            fail("Should not pass, wrote out as\n: "+json);
+        } catch (InvalidDefinitionException e) {
+            verifyException(e, "Java 8 optional type `"+value.getClass().getName()
+                    +"` not supported by default");
+            verifyException(e, "add Module \"com.fasterxml.jackson.datatype:jackson-datatype-jdk8\"");
+        }
+    }
+
+    public void testBetterDeserializationError() throws Exception
+    {
+        _testBetterDeserializationError(Optional.class);
+        _testBetterDeserializationError(OptionalInt.class);
+        _testBetterDeserializationError(OptionalLong.class);
+        _testBetterDeserializationError(OptionalDouble.class);
+    }
+
+    private void _testBetterDeserializationError(Class<?> target) throws Exception
+    {
+        try {
+            Object result = MAPPER.readValue(" 0 ", target);
+            fail("Not expecting to pass, resulted in: "+result);
+        } catch (InvalidDefinitionException e) {
+            verifyException(e, "Java 8 optional type `"+target.getName()+"` not supported by default");
+            verifyException(e, "add Module \"com.fasterxml.jackson.datatype:jackson-datatype-jdk8\"");
+        }
+    }
+
+    // But, [databind#3091], allow deser from JsonToken.VALUE_EMBEDDED_OBJECT
+    public void testAllowAsEmbedded() throws Exception
+    {
+        Optional<Object> optValue = Optional.empty();
+        try (TokenBuffer tb = new TokenBuffer((ObjectCodec) null, false)) {
+            tb.writeEmbeddedObject(optValue);
+
+            try (JsonParser p = tb.asParser()) {
+                Optional<?>  result = MAPPER.readValue(p, Optional.class);
+                assertSame(optValue, result);
+            }
+        }
+
+        // but also try deser into an array
+        try (TokenBuffer tb = new TokenBuffer((ObjectCodec) null, false)) {
+            tb.writeStartArray();
+            tb.writeEmbeddedObject(optValue);
+            tb.writeEndArray();
+
+            try (JsonParser p = tb.asParser()) {
+                Object[] result = MAPPER.readValue(p, Object[].class);
+                assertNotNull(result);
+                assertEquals(1, result.length);
+                assertSame(optValue, result[0]);
+            }
+        }
+    }
+}

